### PR TITLE
Query support for complex types

### PR DESCRIPTION
**Pull request by roji**
Some notes for reviewer and future self:

1. Binding of complex properties and regular properties on complex types
	* `EntityProjectionExpression` (now `StructuralTypeProjectionExpression`) now has `BindComplexProperty`.
	* Unlike regular properties, complex properties - as well as all their nested regular properties - aren't pre-calculated in advance and populated on `StructuralTypeProjectionExpression`. Instead, we generate the correct `StructuralTypeShaperExpression` on-demand, when the complex property is bound.
		* In the future, I intend to change the regular property mechanism to also work on-demand rather than to per-calculate.
	* To support all this, `StructuralTypeProjectionExpression` now also has a *table map*; this maps relational model `ITableBase` to the `TableReferenceExpression` in the current `SelectExpression` which contains the columns for that table.
	* Note that complex property binding happens from `RelationalSqlTranslatingExpressionVisitor` (from `BindComplexProperty`); this is similar to regular properties, but is different from owned navigations which are pre-bound ("expanded") when the lambda body is remapped.
2. Populating the `SelectExpression` projection for entity types which have complex types
	* This is done in `SelectExpression.AddEntityToProjection`, called as part of `ApplyProjection`.
3. Lift projections when pushing down to subquery (`LiftEntityProjectionFromSubquery`)
	* This is very similar to (2) above: it's about populating the subquery's projection with all the nested complex properties.
	* In fact, there seems to be potential for de-duplication between `AddEntityToProjection` (when an entity is projected out of the entire query) and `LiftEntityProjectionFromSubquery` (when an entity is projected out of a subquery). Both populate the `SelectExpression`'s projection.
4. Projecting out a complex type
	* This is for when a complex type is itself projected.
	* Processing lambdas for Select is usually the job of `ProjectionBindingExpressionVisitor`:
		* Lambdas returning a regular property are delegated to `RelationalSqlTranslatingExpressionVisitor`.
		* Lambdas returning an entity type are preprocessed ("expanded") as usual doing remapping, so the entity shaper ends up being the root of the lambda, and handled directly by `ProjectionBindingExpressionVisitor`.
		* Complex types are problematic: we don't expand these during lambda remapping, but rather in `RelationalSqlTranslatingExpressionVisitor`, since we need to bind complex properties in other contexts (e.g. complex type equality).
		* As a result, `RelationalSqlTranslatingExpressionVisitor` now exposes a new `TranslateProjection` API which can return non-SqlExpressions; this is used by `ProjectionBindingExpressionVisitor` to get the complex type shaper expression (which isn't a SqlExpression), similar to how it already handles regular properties.
	* Actual materialization work is is mostly done in `EntityMaterializerSource`. `AddInitializeExpressions` is the thing that generates assignments for regular properties; we now also assign complex properties, calling back into `CreateMaterializeExpression` for the complex type to generate its own nested materializer.
5. Distinct and set operations
	* Complex types are logically the same as any other regular properties directly on the entity type.
	* Doing distinct - or a distinct set operation - for a complex type projection makes all contained properties (recursively) into identifiers.
7. Equality between two complex types
	* With entity type equality we generate equality for the pk properties; for complex types we generate equality for all properties, recursively.
	* Among other things, this requires extending our logic for adding new runtime parameters for allow for recursively extracting properties within complex types, so they can be sent separately as parameters.

Some more ramblings available in #31382

### CODE REVIEW

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 1]
```diff
@@ -1001,8 +1007,8 @@
   <data name="SetOperationsNotAllowedAfterClientEvaluation" xml:space="preserve">
     <value>Unable to translate set operation after client projection has been applied. Consider moving the set operation before the last 'Select' call.</value>
   </data>
-  <data name="SetOperationsOnDifferentStoreTypes" xml:space="preserve">
-    <value>Unable to translate set operation when matching columns on both sides have different store types.</value>
+  <data name="SetOperationOverDifferentStructuralTypes" xml:space="preserve">
+    <value>Set operations over different entity or complex types are not supported ('{type1}' and '{type2}')</value>
```
Comment (ajcvickers) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:1011]: End with a full stop?

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 2]
```diff
@@ -1121,6 +1127,9 @@
   <data name="StoredProcedureUnmapped" xml:space="preserve">
     <value>The entity type '{entityType}' was configured to use some stored procedures and is not mapped to any table. An entity type that isn't mapped to a table must be mapped to insert, update and delete stored procedures.</value>
   </data>
+  <data name="SubqueryOverComplexTypesNotSupported" xml:space="preserve">
+    <value>The query requires a subquery over complex type '{complexType}'; subqueries over complex types are currently unsupported.</value>
```
Comment (ajcvickers) [inline src/EFCore.Relational/Properties/RelationalStrings.resx:1131]: I'm a fan of the semicolon, but to be consistent, this should just be two sentences.

File: src/EFCore/Query/EntityMaterializerSourceParameters.cs [hunk 3]
```diff
@@ -6,36 +6,10 @@ namespace Microsoft.EntityFrameworkCore.Query;
...
-    ///     The query tracking behavior, or <see langword="null" /> if this materialization is not from a query.
-    /// </summary>
-    public QueryTrackingBehavior? QueryTrackingBehavior { get; }
-}
+/// <param name="StructuralType">The entity or complex type being materialized.</param>
+/// <param name="InstanceName">The name of the instance being materialized.</param>
+/// <param name="QueryTrackingBehavior">
+///     The query tracking behavior, or <see langword="null" /> if this materialization is not from a query.
+/// </param>
+public readonly record struct EntityMaterializerSourceParameters(
```
Comment (ajcvickers) [inline src/EFCore/Query/EntityMaterializerSourceParameters.cs:14]: 🥳

**Review by ajcvickers — APPROVED** [review]
:sheep: :it:

File: src/EFCore/Query/EntityMaterializerSourceParameters.cs (hunk 3 above)
Comment (roji) [reply src/EFCore/Query/EntityMaterializerSourceParameters.cs:14]: Right? :)

Am gonna do some MQ work on this :)

File: src/EFCore.Proxies/Proxies/Internal/ProxyBindingInterceptor.cs [hunk 4]
```diff
@@ -38,7 +38,7 @@ public ProxyBindingInterceptor(IProxyFactory proxyFactory)
     /// </summary>
     public virtual InstantiationBinding ModifyBinding(InstantiationBindingInterceptionData interceptionData, InstantiationBinding binding)
     {
-        var entityType = interceptionData.EntityType;
+        var entityType = interceptionData.TypeBase as IEntityType ?? throw new NotImplementedException();
```
Comment (roji) [inline src/EFCore.Proxies/Proxies/Internal/ProxyBindingInterceptor.cs:41]: @ajcvickers I left this one for later - do you feel like taking a look at it?

**Pull request merged**